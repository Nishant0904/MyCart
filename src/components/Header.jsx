import styled from 'styled-components'
import { Link } from 'react-router-dom'
import cartIcon from '../assets/icons/shopping-cart.svg'
import Logo from './reusables/Logo'
import PropTypes from 'prop-types'

const Header = ({ cart }) => {
    return (
        <StyledHeader data-testid='header'>
            <Nav>
                <Logo data-testid='header__logo'/>
                <ul>
                    <li>
                        <NavLink to='/products'>Products</NavLink>
                    </li>
                    <li>
                        <NavLink to='/cart' data-testid='header__cart-link'>
                            <img src={cartIcon} alt='cart icon' />
                            {cart.length > 0 && (
                                <CartItemsNumber>
                                    <div>{cart.length}</div>
                                </CartItemsNumber>
                            )}
                        </NavLink>
                    </li>
                </ul>
            </Nav>
        </StyledHeader>
    )
}

const StyledHeader = styled.header`
    position: sticky;
    top: 0;
    background-color: var(--primary-bg-color);
    z-index: 999;
`

const Nav = styled.nav`
    display: flex;
    align-items: center;
    justify-content: space-between;
    padding: 24px 96px 24px 48px;
    border-bottom: 1px solid black;

    ul {
        display: flex;
        align-items: center;
        gap: 48px;
        list-style: none;
    }

    @media (max-width: 600px) {
        padding: 12px 24px 12px 12px;

        ul {
            gap: 12px;
        }
    }
`

const NavLink = styled(Link)`
    font-size: 0.8rem;
    font-weight: 900;
    text-transform: uppercase;
    letter-spacing: 0.1rem;
    position: relative;
`

const CartItemsNumber = styled.div`
    position: absolute;
    top: -18px;
    right: -16px;
    height: 16px;
    min-width: 16px;
    padding: 4px 2px;
    border-radius: 50%;
    font-size: 0.7rem;
    font-weight: 400;
    display: flex;
    align-items: center;
    justify-content: center;
    background-color: var(--accent-color);
    color: var(--tertiary-txt-color);

    div {
        margin-left: 1px;
    }
`

Header.propTypes = {
    cart: PropTypes.array,
}

export default Header
