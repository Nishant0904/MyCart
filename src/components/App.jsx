import Header from './Header'
import PageContent from './PageContent'
import Footer from './Footer'
import { useState } from 'react'

const App = () => {
    const [cart, setCart] = useState([])

    return (
        <>
            <Header cart={cart} />
            <PageContent
                cart={cart}
                handleAddToCart={handleAddToCart}
                handleDeleteFromCart={handleDeleteFromCart}
            />
            <Footer />
        </>
    )

    function handleAddToCart(newProduct, quantity) {
        if (quantity > 0) {
            let productsToAdd = []
            for (let i = quantity; i > 0; i--) {
                productsToAdd.push(newProduct)
            }
            setCart([...cart, ...productsToAdd])
        }
    }

    function handleDeleteFromCart(currentProduct) {
        const cartWithoutProduct = cart.filter(
            (product) => product.title !== currentProduct.title,
        )
        setCart(cartWithoutProduct)
    }
}

export default App
