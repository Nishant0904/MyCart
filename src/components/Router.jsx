import { createBrowserRouter, RouterProvider } from 'react-router-dom'
import App from './App'
import NotFound from './NotFound'

const Router = () => {
    const router = createBrowserRouter([
        {
            path: '/',
            element: <App />,
            errorElement: <NotFound />,
        },
        {
            path: '/:name',
            element: <App />,
        },
        {
            path: '/products/:id',
            element: <App />,
        },
    ])

    return <RouterProvider router={router} />
}

export default Router
