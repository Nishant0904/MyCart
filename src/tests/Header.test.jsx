import { describe, it, expect } from 'vitest'
import { render, screen, waitFor } from '@testing-library/react'
import Header from '../components/Header'
import Router from '../components/Router'
import userEvent from '@testing-library/user-event'
import App from '../components/App'
import { BrowserRouter } from 'react-router-dom'

describe('Header', () => {
    it('renders header element', () => {
        render(
            <Router>
                <Header />
            </Router>,
        )
        expect(screen.getByTestId('header')).toMatchSnapshot()
    })

    it('Products link navigates to Products page', async () => {
        const user = userEvent.setup()
        render(<App />, { wrapper: BrowserRouter })

        const productsLink = screen.getByRole('link', { name: 'Products' })
        expect(productsLink.href).toContain('/products')

        await user.click(productsLink)
        await waitFor(() => {
            expect(window.location.pathname).toBe('/products')
        })
    })

    it('Cart icon link navigates to cart page', async () => {
        const user = userEvent.setup()
        render(<App />, { wrapper: BrowserRouter })

        const cartLink = screen.getByTestId('header__cart-link')
        expect(cartLink.href).toContain('/cart')

        await user.click(cartLink)
        await waitFor(() => {
            expect(window.location.pathname).toBe('/cart')
        })
    })
})
