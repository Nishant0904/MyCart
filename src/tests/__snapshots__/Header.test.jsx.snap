// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header > renders header element 1`] = `
<header
  class="sc-dmyCSP eTsKeU"
  data-testid="header"
>
  <nav
    class="sc-hLQSwg hSkrVx"
  >
    <a
      class="sc-guDLey dboBSv"
      href="/"
    >
      <img
        alt="cart icon"
        class="sc-beySPh htjDQV"
        src="/src/assets/icons/shopping-cart.svg"
      />
      <p>
        MyCart
      </p>
    </a>
    <ul>
      <li>
        <a
          class="sc-eDLKkx bZsVqJ"
          href="/products"
        >
          Products
        </a>
      </li>
      <li>
        <a
          class="sc-eDLKkx bZsVqJ"
          data-testid="header__cart-link"
          href="/cart"
        >
          <img
            alt="cart icon"
            src="/src/assets/icons/shopping-cart.svg"
          />
        </a>
      </li>
    </ul>
  </nav>
</header>
`;
